sw/banshee/src/main.rs

```diff
@@ -260,6 +293,57 @@ fn main() -> Result<()> {
         .translate_elf(&elf)
         .context("Failed to translate ELF binary")?;
 
+    if has_train_bin {
+        let bin_path = matches.value_of("train-data-bin-file-path").unwrap();
+
+        trace!("Loading train data from binary file: {}", bin_path);
+
+        // get memory offset from argument
+        let mut memory_offset = matches
+            .value_of("train-data-mem-offset")
+            .unwrap()
+            .trim_start_matches("0x");
+        // turn the string into a u64
+        let mut mem_offset = u64::from_str_radix(memory_offset, 16).unwrap();
+
+        trace!("Train data starts at address: 0x{:x}", mem_offset);
+
+        let train_data = dram_preload::bin_read(bin_path, mem_offset).unwrap();
+
+        let train_data_length = train_data.len() as u64;
+
+        let mut mem = engine.memory.lock().unwrap();
+
+        mem.extend(train_data);
+
+        for addr in mem_offset..mem_offset + train_data_length {
+            let val: u32 = mem.get(&(addr)).copied().unwrap_or(0);
+            trace!("address = 0x{:x}, binary value = {:#034b}", addr, val);
+        }
+    }
+
+    if has_train_labels_bin {
+        let bin_path = matches.value_of("train-labels-bin-file-path").unwrap();
```

Comment:
We can get rid of this duplication if we treat preloadable files as a generic list as proposed above. Question is how to treat the different data sizes; they could also be passed similar to the preloading address if necessary.